src/api/telemetry/WebSocketWorker.js

```diff
@@ -0,0 +1,348 @@
+    }
+
+    #message(event) {
+      this.#messageCallbacks.forEach((callback) => callback(event.data));
+    }
+    disconnect() {
+      this.#isConnected = false;
+      this.#isConnecting = false;
+
+      if (this.#webSocket.readyState === WebSocket.OPEN) {
+        this.#webSocket.close();
+      }
+
+      this.dispatchEvent(new Event('disconnected'));
+      this.#webSocket = undefined;
+    }
+
+    #reconnect() {
+      if (this.#reconnectTimeoutHandle) {
+        return;
+      }
+
+      this.#reconnectTimeoutHandle = setTimeout(() => {
+        this.connect(this.#wsUrl);
+
+        this.#reconnectTimeoutHandle = undefined;
+      }, FALLBACK_AND_WAIT_MS[this.#currentWaitIndex]);
+
+      if (this.#currentWaitIndex < FALLBACK_AND_WAIT_MS.length - 1) {
+        this.#currentWaitIndex++;
+      }
+    }
+
+    enqueueMessage(message) {
+      this.#messageQueue.push(message);
+      this.#flushQueueIfReady();
+    }
+
+    #flushQueueIfReady() {
+      if (this.#isConnected) {
+        this.#flushQueue();
+      }
+    }
+
+    #flushQueue() {
+      while (this.#messageQueue.length > 0) {
+        if (!this.#isConnected) {
+          break;
+        }
+
+        const message = this.#messageQueue.shift();
+        this.#webSocket.send(message);
+      }
+    }
+  }
+
+  class WorkerToWebSocketMessageBroker {
+    #websocket;
+    #messageBatcher;
+    #throttledTelemetryEmitter;
+
+    constructor(websocket, messageBatcher, throttledTelemetryEmitter) {
+      this.#websocket = websocket;
+      this.#messageBatcher = messageBatcher;
+      this.#throttledTelemetryEmitter = throttledTelemetryEmitter;
+    }
+
+    routeMessageToHandler(message) {
+      const { type } = message.data;
+      switch (type) {
+        case 'connect':
+          this.connect(message);
+          break;
+        case 'disconnect':
+          this.disconnect(message);
+          break;
+        case 'message':
+          this.#websocket.enqueueMessage(message.data.message);
+          break;
+        case 'setBatchingStrategy':
+          this.setBatchingStrategy(message);
+          break;
+        case 'setRate':
+          this.setRate(message);
+          break;
+        case 'setMaxBatchSize':
+          this.#messageBatcher.setMaxBatchSize(message.data.maxBatchSize);
+          break;
+        default:
+          throw new Error(`Unknown message type: ${type}`);
+      }
+    }
+    connect(message) {
+      const { url } = message.data;
+      this.#websocket.connect(url);
+    }
+    disconnect() {
+      this.#websocket.disconnect();
+    }
+    message(message) {
+      const { subscribeMessage } = message.data;
+      this.#websocket.enqueueMessage(subscribeMessage);
+    }
+    setBatchingStrategy(message) {
+      const { serializedStrategy } = message.data;
+      const batchingStrategy = {
+        // eslint-disable-next-line no-new-func
+        shouldBatchMessage: new Function(`return ${serializedStrategy.shouldBatchMessage}`)(),
```

Comment:
Leaving these warnings for other reviewers to sanity check before I dismiss them. Loathe as I am to be interpreting strings as JavaScript, there is no user-provided value being evaluated here. It is not possible to pass functions across the worker interface without serializing them. It would not be possible to support a customizable batching strategy without allowing plugins to pass custom functions across to the worker. To be clear though, the code being evaluated here is provided by a _developer_ in the form of a plugin, and not a _user_. Therefore it is no riskier than any other _developer_ provided plugin code. I don't see how a user could use this for arbitrary code execution in other users' browsers.

So, I think these two code-injection warnings are false-positives and I'm inclined to dismiss them as such, unless someone can think of a way of otherwise addressing the issue?

Reply:
I guess the idea here is the individual telemetry plugins will provide logic to provide batching strategies, e.g., for YAMCS this is:
```js
this.#socketWorker.setBatchingStrategy({
            shouldBatchMessage: (message) => {
                const type = message.substring(13, message.indexOf("\"", 13));

                return type === 'parameters';
            },
            getBatchIdFromMessage: (message) => {
                const callNumber = message.substring(36, message.indexOf(",", 37));

                return callNumber;
            }
        });
```

and the batching strategies are too complicated to simply pass a regex, or IDs to be extant in the message, etc. If that's the case, this seems like a good approach, and necessary if we want to have this stuff run in a Web Worker.

Reply:
I think a regex would work in this simple case, and it's a good idea actually. It would not work at all with non-string formats like protobuf though, and it would limit developers with more complex use-cases. I'd also be concerned about performance, which was my primary motivation for not using regex in this case.